Ethernet: Add support stm32n6570_dk

marwaiehm-st opened the pull request:
Add ethernet node of stm32n6570_dk
Integrate GRMII and RMII interfaces

Based on this PR: https://github.com/zephyrproject-rtos/zephyr/pull/87593

marwaiehm-st committed:
boards: st: stm32n6570_dk: Add ethernet node

Add the Ethernet MAC and MDIO nodes in the device tree.
Add Kconfig for Ethernet Support.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

marwaiehm-st commented:
west build -p always -b stm32n6570_dk//sb  samples/net/dhcpv4_client/ 

[00:00:00.051,000] <inf> phy_mii: PHY (0) ID 1CC916
*** Booting Zephyr OS build v4.1.0-975-gbe5ce17a90fb ***
[00:00:03.175,000] <inf> net_dhcpv4_client_sample: Run dhcpv4 client
[00:00:03.175,000] <inf> net_dhcpv4_client_sample: Start on ethernet@58036000: index=1
[00:00:06.855,000] <inf> phy_mii: PHY (0) Link speed 1000 Mb, full duplex
[00:00:18.179,000] <inf> net_dhcpv4_client_sample: DHCP Option 42: 192.168.1.1
[00:00:18.191,000] <inf> net_dhcpv4_client_sample: DHCP Option 42: 192.168.1.1
[00:00:18.191,000] <inf> net_dhcpv4: Received: 192.168.1.89
[00:00:18.191,000] <inf> net_dhcpv4_client_sample:    Address[1]: 192.168.1.89
[00:00:18.191,000] <inf> net_dhcpv4_client_sample:     Subnet[1]: 255.255.255.0
[00:00:18.191,000] <inf> net_dhcpv4_client_sample:     Router[1]: 192.168.1.1
[00:00:18.191,000] <inf> net_dhcpv4_client_sample: Lease time[1]: 86400 seconds

maass-hamburg commented:
instead of using and extending the internal ``eth_phy_get_link_state()``, maybe use the ``phy_get_link_state()`` or even ``phy_link_callback_set()`` from the ethernet phy api. The later also has the benefit, that autonegation can be done on every change of the link not just on startup.

maass-hamburg commented:
the ``st,stm32-mdio`` also uses only ``HAL_ETH_ReadPHYRegister()`` and ``HAL_ETH_WritePHYRegister()`` internally, like ``eth_phy_get_link_state()``.  That means, that ``eth_phy_get_link_state()`` can be fully replaced by the use of the ethernet phy api. It just have to make sure, that all soc, that select ETH_STM32_HAL_API_V2 also get the ``st,stm32-mdio`` dev. And all boards the also the generic ``ethernet-phy``.

maass-hamburg commented:
started something like I described there: https://github.com/zephyrproject-rtos/zephyr/pull/87593

marwaiehm-st committed:
drivers: ethernet: Integrate GRMII and RMII interfaces

- Added macros `STM32_ETH_PHY_MODE`
  to determine the PHY mode (RGMII, GMII, RMII, and MII)
  from the `phy_connection_type` property in the device tree.
- Removed previous definitions for ETH_MEDIA_INTERFACE_MII
  and ETH_MEDIA_INTERFACE_RMII.
- Updated STM32_ETH_PHY_MODE macro to use ETH_MII_MODE
  and ETH_RMII_MODE.

Signed-off-by: IBEN EL HADJ MESSAOUD Marwa <marwa.ibenelhadjmessaoud-ext@st.com>

erwango commented on dts/arm/st/f2/stm32f207.dtsi:
```diff
@@ -10,17 +10,28 @@
-				 <&rcc STM32_CLOCK(AHB1, 28U)>;
-			status = "disabled";
+			compatible = "st,stm32-ethernet-parent";
+			clock-names = "stmmaceth";
```
Nit: if this clock is remove from mac, then following would be more suited:
```suggestion
			clock-names = "stm-eth";
```
This would likely require driver update.

maass-hamburg replied:
changed in https://github.com/zephyrproject-rtos/zephyr/pull/87593

erwango replied:
:-D commented in the wrong PR

maass-hamburg commented:
@marwaiehm-st please rebase to current main, now that https://github.com/zephyrproject-rtos/zephyr/pull/87593 is merged

marwaiehm-st commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -80,9 +80,18 @@ static const struct device *eth_stm32_phy_dev = DEVICE_DT_GET(DT_INST_PHANDLE(0,
 #define MAC_NODE DT_NODELABEL(mac)
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)
+#define STM32_ETH_PHY_MODE(node_id) \
+	((DT_ENUM_HAS_VALUE(node_id, phy_connection_type, mii) ? HAL_ETH_MII_MODE : \
+	 (DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rmii) ? HAL_ETH_RMII_MODE : \
+	 (DT_ENUM_HAS_VALUE(node_id, phy_connection_type, gmii) ? HAL_ETH_GMII_MODE : \
+	 (DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rgmii) ? HAL_ETH_RGMII_MODE : \
+		 HAL_ETH_RMII_MODE)))))
```
We used ETH_MEDIA_INTERFACE_FOO to support the api v1, but here we have only stm32n6.
We can add the definition of `ETH_MEDIA_INTERFACE_FOO` for RGMII and GMII, if it makes the code compatible
```c
#define ETH_MEDIA_INTERFACE_RGMII		HAL_ETH_RGMII_MODE
#define ETH_MEDIA_INTERFACE_GMII	        HAL_ETH_GMII_MODE
```

erwango replied:
That would be cleaner indeed

erwango replied:
Or we remove the `ETH_MEDIA_INTERFACE_FOO`, but we should be able to use only one macro version.

marwaiehm-st replied:
We can remove the definition of `ETH_MEDIA_INTERFACE_FOO`  and using directly `HAL_ETH_XMII` for API V2, but for API V1 we can only use `ETH_MEDIA_INTERFACE_FOO `.

In case of removing the definition "ETH_MEDIA_INTERFACE_FOO", we should add definition for HAL_ETH_XMII_MODE to support API v1.
```c
#if !defined(CONFIG_ETH_STM32_HAL_API_V2)
#define HAL_ETH_MII_MODE   ETH_MEDIA_INTERFACE_MII
#define HAL_ETH_RMII_MODE ETH_MEDIA_INTERFACE_RMII
#endif

#define STM32_ETH_PHY_MODE(node_id) \
	(DT_ENUM_HAS_VALUE(node_id, phy_connection_type, mii) ? \
		HAL_ETH_MII_MODE : HAL_ETH_RMII_MODE)
```

maass-hamburg commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1477,7 +1493,9 @@ static const struct eth_stm32_hal_dev_cfg eth0_config = {
 BUILD_ASSERT(DT_ENUM_HAS_VALUE(MAC_NODE, phy_connection_type, mii) ||
-	     DT_ENUM_HAS_VALUE(MAC_NODE, phy_connection_type, rmii),
+	     DT_ENUM_HAS_VALUE(MAC_NODE, phy_connection_type, rmii) ||
+	     DT_ENUM_HAS_VALUE(MAC_NODE, phy_connection_type, rgmii) ||
```
maybe use a ``IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet),(check for rgmii and gmii))``, to only allow gigabit connection types on supported soc

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -1138,7 +1149,12 @@ static void set_mac_config(const struct device *dev, struct phy_link_state *stat
 
-	mac_config.Speed = PHY_LINK_IS_SPEED_100M(state->speed) ? ETH_SPEED_100M : ETH_SPEED_10M;
+	mac_config.Speed =
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)
+		PHY_LINK_IS_SPEED_1000M(state->speed) ? ETH_SPEED_1000M :
+#endif
```
Can `IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)` could also be used here ?

erwango commented on drivers/ethernet/eth_stm32_hal.c:
```diff
@@ -78,11 +75,33 @@ static const struct device *eth_stm32_phy_dev = DEVICE_DT_GET(DT_INST_PHANDLE(0,
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32n6_ethernet)
+#define STM32_ETH_PHY_MODE(node_id) \
+	((DT_ENUM_HAS_VALUE(node_id, phy_connection_type, mii) ? ETH_MII_MODE : \
+	 (DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rmii) ? ETH_RMII_MODE : \
+	 (DT_ENUM_HAS_VALUE(node_id, phy_connection_type, gmii) ? ETH_GMII_MODE : \
+	 (DT_ENUM_HAS_VALUE(node_id, phy_connection_type, rgmii) ? ETH_RGMII_MODE : \
+		ETH_RMII_MODE)))))
```
Since property is required, there is no need to default to `ETH_RMII_MODE` while there's already a check on `ETH_RMII_MODE`

marwaiehm-st replied:
Here, we must have a default, so I deleted the check on RMII and kept the default.
